### PR TITLE
Fix removed invisible item children

### LINKED ISSUE #542: In item edition, children edition: fix: non visible children get removed at save
## Overview

For the moment, we get the list of children, allow the user to update it and then save the whole list of children. The problem is that we might have some children which are not visible by the current user, and as they are not returned to UI, on save they are being removed from the child list at save.

## Acceptance criteria
- [x] Use the new "show_invisible_items" parameter on the "itemChildrenView" service to receive to include the items which are not visible by the current user. That might require to create a different function in the service which returns a different type, so that most attributes do not become optional for other usage of this service.
- [x] If the item is not visible, just display something like (in italic or grey or ?) "Non-visible item (`<id>`)" (without link) to allow the re-ordering and removal of the item if needed.
- [x] make sure the other usages of the service are not impacted (the list of children for chapters/skills should not display invisible items)
- [x] test it properly, if you need some data to be created to test, let me know

### PR DESCRIPTION
**Pull request by ThomasJuster**
Fixes #542

### CODE REVIEW

**Review by smadbe — CHANGES_REQUESTED** [review]

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.ts [hunk 1]
```diff
@@ -172,7 +173,7 @@ export class ItemChildrenEditComponent implements OnChanges {
     }
 
     this.itemRouter.navigateTo({
-      contentType: typeCategoryOfItem(child),
+      contentType: typeCategoryOfItem(child as { type: ItemType }),
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.ts]: Fixed

File: src/app/modules/item/components/item-children-edit/item-children-edit.component.ts [hunk 2]
```diff
@@ -7,14 +7,15 @@ import { ItemType, typeCategoryOfItem } from '../../../../shared/helpers/item-ty
...
+import { isVisibleChild } from '../../helpers/item-permissions';
 import { isNotUndefined } from '../../../../shared/helpers/null-undefined-predicates';
 
 export interface ChildData {
   id?: string,
   title: string | null,
-  type: ItemType,
+  type: ItemType | null,
   scoreWeight: number,
+  isVisible: boolean;
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/components/item-children-edit/item-children-edit.component.ts]: Fixed

File: src/app/modules/item/helpers/item-permissions.ts [hunk 3]
```diff
@@ -15,3 +16,7 @@ export interface ItemPermissionsInfo {
 export function canCurrentUserViewItemContent(item: ItemPermissionsInfo): boolean {
   return [ 'content','content_with_descendants','solution' ].includes(item.permissions.canView);
 }
+
+export function isVisibleChild(item: ItemPermissionsInfo): item is ItemVisibleChild {
+  return item.permissions.canView !== 'none';
+}
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/helpers/item-permissions.ts]: Fixed

File: src/app/modules/item/http-services/get-item-children.service.ts [hunk 4]
```diff
@@ -32,7 +20,33 @@ export const itemChildDecoder = D.struct({
   watchPropagation: D.boolean,
 });
 
-export type ItemChild = D.TypeOf<typeof itemChildDecoder>;
+const itemVisibleChildDecoder = D.intersect(baseItemChildDecoder)(
```
[the remark this answers is not in the document]
Comment (ThomasJuster) [reply src/app/modules/item/http-services/get-item-children.service.ts]: Fixed

**Comment by ThomasJuster** [comment]
To be tested on item 561969155071897034:
[localhost](http://localhost:4200/#/activities/by-id/561969155071897034;parentAttempId=0;path=4702,1352246428241737349/edit)
[dev.algorea.org](https://dev.algorea.org/branch/fix/removed-invisible-item-children/en/#/activities/by-id/561969155071897034;parentAttempId=0;path=4702,1352246428241737349/edit)

**Review by smadbe — APPROVED** [review]
Everything looks good.

**Pull request merged**